p2p/dandelion/src/router.rs

```diff
@@ -0,0 +1,384 @@
+            },
+            TxState::Local => {
+                tracing::debug!(parent: &self.span, "Steming local tx.");
+
+                Box::pin(
+                    self.stem_local_tx(req.tx)
+                        .map_ok(|_| State::Stem)
+                        .map_err(DandelionRouterError::PeerError),
+                )
+            }
+        }
+    }
+}
+
+/// A [`DandelionRouter`] builder.
+///
+/// You _must_ call all functions, no defaults are set.
+#[derive(Debug)]
+pub struct DandelionRouterBuilder<P, B, Tx> {
+    config: Option<DandelionConfig>,
+    /// A [`Discover`] where we can get outbound peers from.
+    outbound_peer_discover: Option<P>,
+    /// A [`Service`] which handle broadcasting (diffusing) transactions.
+    broadcast_svc: Option<B>,
+
+    _tx: PhantomData<Tx>,
+}
+
+impl<P, B, Tx> Default for DandelionRouterBuilder<P, B, Tx> {
+    fn default() -> Self {
+        Self {
+            config: None,
+            outbound_peer_discover: None,
+            broadcast_svc: None,
+
+            _tx: PhantomData,
+        }
+    }
+}
+
+impl<ID, S, P, B, Tx> DandelionRouterBuilder<P, B, Tx>
+where
+    ID: Hash + Eq + Clone,
+    P: Discover<Key = ID, Service = S, Error = tower::BoxError>,
+    B: Service<DiffuseRequest<Tx>, Error = tower::BoxError>,
+    B::Future: Send + 'static,
+    S: Service<StemRequest<Tx>, Error = tower::BoxError>,
+    S::Future: Send + 'static,
+{
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn with_config(self, config: DandelionConfig) -> Self {
+        Self {
+            config: Some(config),
+            ..self
+        }
+    }
+
+    pub fn with_outbound_peer_discover(self, outbound_peer_discover: P) -> Self {
+        Self {
+            outbound_peer_discover: Some(outbound_peer_discover),
+            ..self
+        }
+    }
+
+    pub fn with_broadcast_svc(self, broadcast_svc: B) -> Self {
+        Self {
+            broadcast_svc: Some(broadcast_svc),
+            ..self
+        }
+    }
+
+    pub fn build(self) -> DandelionRouter<P, B, ID, S, Tx> {
+        let config = self.config.unwrap();
+
+        // get the current state
+        let state_dist = Bernoulli::new(config.fluff_probability)
+            .expect("Fluff probability was not between 0 and 1");
+
+        let current_state = if state_dist.sample(&mut thread_rng()) {
+            State::Fluff
+        } else {
+            State::Stem
+        };
+
+        DandelionRouter {
+            outbound_peer_discover: Box::pin(self.outbound_peer_discover.unwrap()),
+            broadcast_svc: self.broadcast_svc.unwrap(),
+            current_state,
+            epoch_start: Instant::now(),
+            local_route: None,
+            stem_routes: HashMap::new(),
+            stem_peers: HashMap::new(),
+            state_dist,
+            config,
+            span: tracing::debug_span!("dandelion_router", state = ?current_state),
+            _tx: PhantomData,
+        }
+    }
```

Comment:
If every field must be set I don't think it needs to be/is a builder.

This could just take the values as input in a constructor function directly instead of panics? If builder style is still wanted, a panic doc would be nice.

Reply:
Yeah I thought there were more fields than there turned out to be which is why I created a builder, seen as there are only 3 I'll remove the builder